modules/repeater/src/main/scala/com.snowplowanalytics.snowplow.storage.bigquery.repeater/Resources.scala

```diff
@@ -69,34 +72,36 @@ final class Resources[F[_]: Sync](
       newLifetime = Duration(now.getMillis - jobStartTime.getMillis, "millis")
       _ <- statistics.update(s => s.copy(lifetime = newLifetime))
     } yield ()
-
   def showStats(implicit L: Logger[F]): F[Unit] =
     statistics.get.flatMap(statistics => L.info(statistics.show))
 }
 
 object Resources {
   private val QueueSize = 100
 
-  final case class Statistics(inserted: Int, uninsertable: Int, lifetime: Duration)
+  final case class Statistics(inserted: Int, uninsertable: Int, retries: Int, lifetime: Duration)
   object Statistics {
-    val start: Statistics = Statistics(0, 0, Duration(0, "millis"))
+    val start: Statistics = Statistics(0, 0, 0, Duration(0, "millis"))
 
     implicit val showRepeaterStatistics: Show[Statistics] = {
       case s if (s.lifetime.toHours == 0) =>
-        s"Statistics: ${s.inserted} rows inserted, ${s.uninsertable} rows rejected in ${s.lifetime.toMinutes} minutes."
+        s"${common(s)} ${s.lifetime.toMinutes} minutes."
       case s if (s.lifetime.toHours > 24) =>
-        s"Statistics: ${s.inserted} rows inserted, ${s.uninsertable} rows rejected in ${s.lifetime.toDays} days and ${s.lifetime.toHours - s.lifetime.toDays * 24} hours."
+        s"${common(s)} ${s.lifetime.toDays} days and ${s.lifetime.toHours - s.lifetime.toDays * 24} hours."
       case s =>
-        s"Statistics: ${s.inserted} rows inserted, ${s.uninsertable} rows rejected in ${s.lifetime.toHours} hours."
+        s"${common(s)} ${s.lifetime.toHours} hours."
     }
+
+    private def common(s: Statistics): String =
+      s"Statistics: ${s.inserted} rows inserted, ${s.uninsertable} rows rejected in"
   }
 
   /** Allocate all resources */
-  def acquire[F[_]: ConcurrentEffect: Timer: Logger](
+  def acquire[F[_]: ConcurrentEffect: ContextShift: Timer: Logger](
```

Comment:
It is very unusual to see code explicitly creating a blocker from a thread pool unless there is a very good reason.  Normally people use the `Blocker[F]` constructor, and indeed that is what you do in [the streamloader module](https://github.com/snowplow-incubator/snowplow-bigquery-loader/blob/1.0.0-rc5/modules/streamloader/src/main/scala/com/snowplowanalytics/snowplow/storage/bigquery/streamloader/Resources.scala#L69).

Personally I would re-write this function as a single for loop, instead of getting mixed up in intermediate types like `F[Blocker => F[Resources[F]]]`.

This compiles and seems to work.

```scala
  def acquire[F[_]: ConcurrentEffect: ContextShift: Timer: Logger](
    cmd: RepeaterCli.ListenCommand
  ): Resource[F, Resources[F]] =
    Blocker[F].evalMap { blocker =>
      val env = cmd.env
      for {
        bigQuery <- services.Database.getClient[F]
        bucket <- Sync[F].fromEither(
          validateBucket(env.config.output.deadLetters.bucket)
            .toEither
            .leftMap(e => new RuntimeException(e.toList.mkString("\n")))
        )
        queue       <- Queue.bounded[F, BadRow](QueueSize)
        counter     <- Ref[F].of[Int](0)
        stop        <- SignallingRef[F, Boolean](false)
        statistics  <- Ref[F].of[Statistics](Statistics.start)
        concurrency <- Sync[F].delay(Runtime.getRuntime.availableProcessors * 16)
        _ <- Logger[F].info(
          s"Initializing Repeater from ${env.config.input.subscription} to ${env.config.output.good} with $concurrency streams"
        )
        jobStartTime <- Sync[F].delay(Instant.now())
        reporter     <- mkMetricsReporter[F](blocker, env.monitoring)
      } yield new Resources[F](
        bigQuery,
        bucket,
        env,
        queue,
        counter,
        stop,
        statistics,
        cmd.bufferSize,
        cmd.timeout,
        cmd.backoffPeriod,
        concurrency,
        blocker,
        jobStartTime,
        reporter
      )
    }

```

Reply:
https://github.com/snowplow-incubator/snowplow-bigquery-loader/issues/216